crates/data-plane-controller/src/controller.rs

```diff
@@ -157,16 +157,7 @@ impl Controller {
             self.validate_state(ops_checkout, state).await?;
         }
 
-        if !state.private_links.is_empty()
-            && !state.stack.config.model.private_links.is_empty()
-            && state.private_links != state.stack.config.model.private_links
-        {
-            anyhow::bail!(
-                "cannot set both config.private_links and private_links, prefer the latter."
-            );
-        }
-
-        state.stack.config.model.private_links = state.private_links.clone();
+        state.stack.config.model.private_links = row_state.stack.config.model.private_links.clone();
```

Comment:
this certainly helps. :thinking: the problem we'll run into, though, is that on_idle will be unable to detect that  private link config has changed (because it's already overwritten at that point), and thus that it needs to kick off a convergence pass.

Perhaps instead we should thread in `row_state`'s private links into `on_pulumi_up_1` ? Then, that routine could then update `state.stack.config.model.private_links` after they've been applied and are up.

Reply:
How about replacing the line you have commented on (160) with:

```rust
        if row_state.stack.config.model.private_links != state.stack.config.model.private_links {
            state.stack.config.model.private_links =
                row_state.stack.config.model.private_links.clone();
            state.pending_converge = true;
        }
```

Reply:
That could work. The difference is we'd do an extra convergence if a) on_idle detects an initial (broken) config update, b) after attempted `pulumi up`'s, the config is fixed by an operator, c) the updated private_link config is applied, we successfully converge, and then d) we do an extra converge because `pending_converge = true`.

I have a mild preference for threading in row_state's private_links into on_pulumi_up_1, and updating the stack state at that time, because it 1) precisely captures an updated/fixed config at the time it's effectively applied and 2) it holds those private_links constant (now that we've confirmed it's working) even if the DB column is changed during `on_pulumi_up_2`, which seems desireable.